arangod/Agency/Agent.cpp

```diff
@@ -1155,15 +1182,49 @@ void Agent::run() {
       _agentNeedsWakeup = false;
     }
 
+    if (leading() && getPrepareLeadership() == 1) {
```

Comment:
Looks like race condition:  leading() is looking to constituent leading() which requires _castLock.  But Constituent::lead() releases the _castLock before setting BeginPrepareLeadership().  So there is a moment when it is possible for leading() to be true and getPrepareLeadership() to still be zero (on the way to becoming 1).

Reply:
suggest moving beginPrepareLeadership() within _castLock's block.  And signal agent's _appendCV in a subsequent block outside _castLock's block.

Reply:
Good catch again, I followed your suggestion.